File: bento/src/test/java/com/yelp/android/bento/components/ListComponentTest.java
package com.yelp.android.bento.components;

import static junit.framework.Assert.assertEquals;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.verify;

import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.GridLayoutManager.SpanSizeLookup;
import java.util.ArrayList;
import java.util.List;
import org.junit.Test;

public class ListComponentTest {

    private ListComponent<Void, Object> mListComponent;

    @Test
    public void removeItemGettersFiveItemsFirstItemNoDividers_FirstItemCountOne() {
        setup(5);
        mListComponent.toggleDivider(false);

        assertEquals(
                "Expected to start removing at the first item",
                0,
                mListComponent.getRemoveIndexStart(0));
        assertEquals("Only expected to remove one item", 1, mListComponent.getRemoveItemCount());
    }

    @Test
    public void removeItemGettersFiveItemsMiddleItemNoDividers_MiddleItemCountOne() {
        setup(5);
        mListComponent.toggleDivider(false);

        assertEquals(
                "Expected to start removing at the middle item",
                2,
                mListComponent.getRemoveIndexStart(2));
        assertEquals("Only expected to remove one item", 1, mListComponent.getRemoveItemCount());
    }

    @Test
    public void removeItemGettersFiveItemsLastItemNoDividers_LastItemCountOne() {
        setup(5);
        mListComponent.toggleDivider(false);

        assertEquals(
                "Expected to start removing at the Last item",
                4,
                mListComponent.getRemoveIndexStart(4));
        assertEquals("Only expected to remove one item", 1, mListComponent.getRemoveItemCount());
    }

    @Test
    public void removeItemGettersSingleItemOnlyItemNoDividers_OnlyItemCountOne() {
        setup(1);
        mListComponent.toggleDivider(false);

        assertEquals(
                "Expected to start removing at the only item",
                0,
                mListComponent.getRemoveIndexStart(0));
        assertEquals("Only expected to remove one item", 1, mListComponent.getRemoveItemCount());
    }

    @Test
    public void removeItemGettersFiveItemsFirstItemWithDividers_FirstItemCountTwo() {
        setup(5);
        mListComponent.toggleDivider(true);

        assertEquals(
                "Expected to start removing at the first item",
                0,
                mListComponent.getRemoveIndexStart(0));
        assertEquals("Expected to remove two items", 2, mListComponent.getRemoveItemCount());
    }

    @Test
    public void removeItemGettersFiveItemsMiddleItemWithDividers_MiddleItemTimesTwoMinusOneCountTwo() {
        setup(5);
        mListComponent.toggleDivider(true);

        assertEquals(
                "Expected to start removing at the first item",
                3,
                mListComponent.getRemoveIndexStart(2));
        assertEquals("Expected to remove two items", 2, mListComponent.getRemoveItemCount());
    }

    @Test
    public void removeItemGettersFiveItemsLastItemWithDividers_LastItemTimesTwoMinusOneCountTwo() {
        setup(5);
        mListComponent.toggleDivider(true);

        assertEquals(
                "Expected to start removing at the first item",
                7,
                mListComponent.getRemoveIndexStart(4));
        assertEquals("Expected to remove two items", 2, mListComponent.getRemoveItemCount());
    }

    @Test
    public void removeItemGettersSingleItemOnlyItemWithDividers_FirstItemCountOne() {
        setup(1);
        mListComponent.toggleDivider(true);

        assertEquals(
                "Expected to start removing at the first item",
                0,
                mListComponent.getRemoveIndexStart(0));
        assertEquals("Expected to remove one item", 1, mListComponent.getRemoveItemCount());
    }

    @Test
    public void getSpanSizeLookupTwoColumnsGapItem_ReturnsTwo() {
        setup(2, 2);
        mListComponent.setStartGap(10);
        GridLayoutManager.SpanSizeLookup spanSizeLookup = mListComponent.getSpanSizeLookup();
        assertEquals(2, spanSizeLookup.getSpanSize(0));
    }

    @Test
    public void getSpanSizeLookupTwoColumnsNonGapItem_ReturnsOne() {
        setup(2, 2);
        mListComponent.setStartGap(10);
        GridLayoutManager.SpanSizeLookup spanSizeLookup = mListComponent.getSpanSizeLookup();
        assertEquals(1, spanSizeLookup.getSpanSize(1));
    }

    @Test
    public void setSpanSizeLookupCustomReturnsTen_ReturnsLookupWithTen() {
        setup(2);
        SpanSizeLookup customLookup = new SpanSizeLookup() {
            @Override
            public int getSpanSize(int position) {
                return 10;
            }
        };

        mListComponent.setSpanSizeLookup(customLookup);

        assertEquals(10, mListComponent.getSpanSizeLookup().getSpanSize(0));
    }

    @Test
    public void whenDividerEnabledAndOnItemVisibleCalled_CallsOnListItemVisibleForCorrectIndices() {
        int numberOfItems = 10;
        setup(numberOfItems);

        ListComponent spy = spy(mListComponent);
        for (int i = 0; i < numberOfItems; i++) {
            spy.onItemVisible(i);
            // When the index is an odd number, onListItemVisible() should still have been called
            // only once because the item in an even number index is a divider and we should no-op
            // for dividers.
            verify(spy).onListItemVisible(i / 2);
            if (i > numberOfItems / 2) {
                verify(spy, never()).onListItemVisible(i);
            }
        }
    }

    @Test
    public void whenDividerDisabledAndOnItemVisibleCalled_CallsOnListItemVisibleForCorrectIndices() {
        int numberOfItems = 10;
        setup(numberOfItems);
        mListComponent.toggleDivider(false);

        ListComponent spy = spy(mListComponent);
        for (int i = 0; i < numberOfItems; i++) {
            spy.onItemVisible(i);
            verify(spy).onListItemVisible(i);
        }
    }

    @Test
    public void whenDividerEnabledAndOnItemNotVisibleCalled_CallsOnListItemNotVisibleForCorrectIndices() {
        int numberOfItems = 10;
        setup(numberOfItems);

        ListComponent spy = spy(mListComponent);
        for (int i = 0; i < numberOfItems; i++) {
            spy.onItemNotVisible(i);
            // When the index is an odd number, onListItemNotVisible() should still have been called
            // only once because the item in an even number index is a divider and we should no-op
            // for dividers.
            verify(spy).onListItemNotVisible(i / 2);
            if (i > numberOfItems / 2) {
                verify(spy, never()).onListItemNotVisible(i);
            }
        }
    }

    @Test
    public void whenDividerDisabledAndOnItemNotVisibleCalled_CallsOnListItemNotVisibleForCorrectIndices() {
        int numberOfItems = 10;
        setup(numberOfItems);
        mListComponent.toggleDivider(false);

        ListComponent spy = spy(mListComponent);
        for (int i = 0; i < numberOfItems; i++) {
            spy.onItemNotVisible(i);
            verify(spy).onListItemNotVisible(i);
        }
    }

    /**
     * Adds fake items to the list component.
     *
     * @param count The number of items to add.
     * @param lanes The number of columns the list should have
     */
    private void setup(int count, int lanes) {
        mListComponent = new ListComponent<>(null, null, lanes);
        List<Object> fakeData = new ArrayList<>(count);

        while (count > 0) {
            count--;
            fakeData.add(new Object());
        }
        mListComponent.appendData(fakeData);
    }

    private void setup(int count) {
        setup(count, 1);
    }
}

File: bento/src/test/java/com/yelp/android/bento/components/ListComponentTest.kt
package com.yelp.android.bento.components

import androidx.recyclerview.widget.GridLayoutManager.SpanSizeLookup
import com.nhaarman.mockitokotlin2.never
import com.nhaarman.mockitokotlin2.spy
import com.nhaarman.mockitokotlin2.verify
import com.yelp.android.bento.core.TestComponentViewHolder
import org.junit.Assert.assertEquals
import org.junit.Test

class ListComponentTest {

    private lateinit var listComponent: ListComponent<Nothing?, Nothing?>

    @Test
    fun removeItemGettersFiveItemsFirstItemNoDividers_FirstItemCountOne() {
        setup(5)
        listComponent.toggleDivider(false)

        assertEquals(
                "Expected to start removing at the first item",
                0,
                listComponent.getRemoveIndexStart(0))
        assertEquals("Only expected to remove one item", 1, listComponent.removeItemCount)
    }

    @Test
    fun removeItemGettersFiveItemsMiddleItemNoDividers_MiddleItemCountOne() {
        setup(5)
        listComponent.toggleDivider(false)

        assertEquals(
                "Expected to start removing at the middle item",
                2,
                listComponent.getRemoveIndexStart(2))
        assertEquals("Only expected to remove one item", 1, listComponent.removeItemCount)
    }

    @Test
    fun removeItemGettersFiveItemsLastItemNoDividers_LastItemCountOne() {
        setup(5)
        listComponent.toggleDivider(false)

        assertEquals(
                "Expected to start removing at the Last item",
                4,
                listComponent.getRemoveIndexStart(4))
        assertEquals("Only expected to remove one item", 1, listComponent.removeItemCount)
    }

    @Test
    fun removeItemGettersSingleItemOnlyItemNoDividers_OnlyItemCountOne() {
        setup(1)
        listComponent.toggleDivider(false)

        assertEquals(
                "Expected to start removing at the only item",
                0,
                listComponent.getRemoveIndexStart(0))
        assertEquals("Only expected to remove one item", 1, listComponent.removeItemCount)
    }

    @Test
    fun removeItemGettersFiveItemsFirstItemWithDividers_FirstItemCountTwo() {
        setup(5)
        listComponent.toggleDivider(true)

        assertEquals(
                "Expected to start removing at the first item",
                0,
                listComponent.getRemoveIndexStart(0))
        assertEquals("Expected to remove two items", 2, listComponent.removeItemCount)
    }

    @Test
    fun removeItemGettersFiveItemsMiddleItemWithDividers_MiddleItemTimesTwoMinusOneCountTwo() {
        setup(5)
        listComponent.toggleDivider(true)

        assertEquals(
                "Expected to start removing at the first item",
                3,
                listComponent.getRemoveIndexStart(2))
        assertEquals("Expected to remove two items", 2, listComponent.removeItemCount)
    }

    @Test
    fun removeItemGettersFiveItemsLastItemWithDividers_LastItemTimesTwoMinusOneCountTwo() {
        setup(5)
        listComponent.toggleDivider(true)

        assertEquals(
                "Expected to start removing at the first item",
                7,
                listComponent.getRemoveIndexStart(4))
        assertEquals("Expected to remove two items", 2, listComponent.removeItemCount)
    }

    @Test
    fun removeItemGettersSingleItemOnlyItemWithDividers_FirstItemCountOne() {
        setup(1)
        listComponent.toggleDivider(true)

        assertEquals(
                "Expected to start removing at the first item",
                0,
                listComponent.getRemoveIndexStart(0))
        assertEquals("Expected to remove one item", 1, listComponent.removeItemCount)
    }

    @Test
    fun getSpanSizeLookupTwoColumnsGapItem_ReturnsTwo() {
        setup(2, 2)
        listComponent.setStartGap(10)
        val spanSizeLookup = listComponent.spanSizeLookup
        assertEquals(2, spanSizeLookup.getSpanSize(0))
    }

    @Test
    fun getSpanSizeLookupTwoColumnsNonGapItem_ReturnsOne() {
        setup(2, 2)
        listComponent.setStartGap(10)
        val spanSizeLookup = listComponent.spanSizeLookup
        assertEquals(1, spanSizeLookup.getSpanSize(1))
    }

    @Test
    fun setSpanSizeLookupCustomReturnsTen_ReturnsLookupWithTen() {
        setup(2)
        listComponent.spanSizeLookup = object : SpanSizeLookup() {
            override fun getSpanSize(position: Int): Int {
                return 10
            }
        }

        assertEquals(10, listComponent.spanSizeLookup.getSpanSize(0))
    }

    @Test
    fun whenDividerEnabledAndOnItemVisibleCalled_CallsOnListItemVisibleForCorrectIndices() {
        val numberOfItems = 10
        setup(numberOfItems)

        val spy = spy(listComponent)
        for (i in 0 until numberOfItems) {
            spy.onItemVisible(i)
            // When the index is an odd number, onListItemVisible() should still have been called
            // only once because the item in an even number index is a divider and we should no-op
            // for dividers.
            verify(spy).onListItemVisible(i / 2)
            if (i > numberOfItems / 2) {
                verify(spy, never()).onListItemVisible(i)
            }
        }
    }

    @Test
    fun whenDividerDisabledAndOnItemVisibleCalled_CallsOnListItemVisibleForCorrectIndices() {
        val numberOfItems = 10
        setup(numberOfItems)
        listComponent.toggleDivider(false)

        val spy = spy(listComponent)
        for (i in 0 until numberOfItems) {
            spy.onItemVisible(i)
            verify(spy).onListItemVisible(i)
        }
    }

    @Test
    fun whenDividerEnabledAndOnItemNotVisibleCalled_CallsOnListItemNotVisibleForCorrectIndices() {
        val numberOfItems = 10
        setup(numberOfItems)

        val spy = spy(listComponent)
        for (i in 0 until numberOfItems) {
            spy.onItemNotVisible(i)
            // When the index is an odd number, onListItemNotVisible() should still have been called
            // only once because the item in an even number index is a divider and we should no-op
            // for dividers.
            verify(spy).onListItemNotVisible(i / 2)
            if (i > numberOfItems / 2) {
                verify(spy, never()).onListItemNotVisible(i)
            }
        }
    }

    @Test
    fun whenDividerDisabledAndOnItemNotVisibleCalled_CallsOnListItemNotVisibleForCorrectIndices() {
        val numberOfItems = 10
        setup(numberOfItems)
        listComponent.toggleDivider(false)

        val spy = spy(listComponent)
        for (i in 0 until numberOfItems) {
            spy.onItemNotVisible(i)
            verify(spy).onListItemNotVisible(i)
        }
    }

    @Test
    fun appendingDataWithDividers_NotifiesCorrectIndices() {
        setup(5)
        listComponent.toggleDivider(true)

        val listSpy = spy(listComponent)
        listSpy.appendData(listOf(null, null, null))
        // We started with 5 items and inserted 3. Since there are dividers, we need to offset both
        // the start index and the number being added (since we will also add dividers for the new
        // items)
        verify(listSpy).notifyItemRangeInserted(9, 6)
    }

    @Test
    fun appendingDataWithoutDividers_NotifiesCorrectIndices() {
        setup(5)
        listComponent.toggleDivider(false)

        val listSpy = spy(listComponent)
        listSpy.appendData(listOf(null, null, null))
        // We started with 5 items and inserted 3.
        verify(listSpy).notifyItemRangeInserted(5, 3)
    }

    /**
     * Adds fake items to the list component.
     *
     * @param count The number of items to add.
     * @param lanes The number of columns the list should have
     */
    private fun setup(count: Int, lanes: Int = 1) {
        listComponent = ListComponent(null, TestComponentViewHolder::class.java, lanes)
        listComponent.setData((1..count).map { null })
    }
}
